File: AI_data_tool/data_analytics/frontend/src/lib/semanticGuard.ts
/**
 * The semantic veto (MASTER_PLAN Phase 0, constitution rule 3): a column can be
 * numeric without being a QUANTITY. Summing latitudes, years or customer ids
 * returns a number, every time, with no error -- and the number means nothing.
 * SAS defaults every numeric to Sum and its suggestion engine proposes summed
 * latitudes; about a third of its catalogued defects trace back to this.
 *
 * Classified by NAME, the way these columns are actually named, and mirroring
 * the backend's own rules so both halves agree: `_ID_NAME_PATTERN` in
 * services/widget_data.py, and the lat/lon words in services/dataset_profile.py.
 */
export type NonAdditiveKind = 'coordinate' | 'identifier' | 'year'

const ID_NAME = /(^id$|_id$|_key$|_uuid$|^uuid$|_code$)/i
const WORD = (name: string, words: string[]) => {
  const low = name.toLowerCase()
  return words.some(w => low === w || low.endsWith('_' + w) || low.startsWith(w + '_') || low.includes('_' + w + '_'))
}

export function nonAdditiveKind(column: string | undefined | null): NonAdditiveKind | null {
  if (!column) return null
  if (WORD(column, ['lat', 'latitude', 'lon', 'lng', 'long', 'longitude'])) return 'coordinate'
  if (ID_NAME.test(column)) return 'identifier'
  if (WORD(column, ['year', 'yr', 'fiscal_year'])) return 'year'
  return null
}

/** The aggregation that means something for each kind -- the one-click fix. */
export const SAFE_AGGREGATION: Record<NonAdditiveKind, { value: string; label: string }> = {
  coordinate: { value: 'avg', label: 'Average' },
  identifier: { value: 'countd', label: 'Distinct count' },
  year:       { value: 'max', label: 'Maximum' },
}

/** Aggregations that add values up (or average them) -- meaningless for these kinds. */
const ARITHMETIC: Record<NonAdditiveKind, Set<string>> = {
  coordinate: new Set(['sum']),
  identifier: new Set(['sum', 'avg', 'mean', 'average', 'median', 'stddev', 'variance']),
  year:       new Set(['sum', 'avg', 'mean', 'average']),
}

/** A warning for this (column, aggregation), or null when it is meaningful. */
export function semanticAggregationWarning(column: string | undefined | null, aggregation: string): string | null {
  const kind = nonAdditiveKind(column)
  if (!kind || !ARITHMETIC[kind].has((aggregation || '').toLowerCase())) return null
  const what = aggregation.toLowerCase() === 'sum' ? 'Summing' : 'Averaging'
  if (kind === 'coordinate') return `${what} ${column} adds up map coordinates — the result is not a place. Use Average for a centre point, or draw it on a map.`
  if (kind === 'identifier') return `${what} ${column} does arithmetic on identifiers — the result means nothing. Count distinct ${column} instead.`
  return `${what} ${column} adds up years — the result is not a year. Group by it, or take its Maximum.`
}

File: AI_data_tool/data_analytics/backend/app/services/semantic_guard.py
"""The semantic veto, backend half (MASTER_PLAN Phase 0, constitution rule 3).

A column can be numeric without being a quantity. Summing, averaging,
correlating or trending a latitude, an identifier or a year returns a number
every time, with no error, and the number means nothing -- the suggestion
engine SAS ships proposes summed latitudes for exactly this reason.

Classified by NAME, the way these columns are actually named. Mirrors
frontend/src/lib/semanticGuard.ts; `tests/test_semantic_guard.py` pins the two
to the same examples so they cannot drift. Pure: no pandas, no FastAPI.
"""
from __future__ import annotations

import re

_ID_NAME = re.compile(r"(^id$|_id$|_key$|_uuid$|^uuid$|_code$)", re.I)
_COORD_WORDS = ("lat", "latitude", "lon", "lng", "long", "longitude")
_YEAR_WORDS = ("year", "yr", "fiscal_year")


def _word(name: str, words) -> bool:
    low = str(name).lower()
    return any(low == w or low.endswith("_" + w) or low.startswith(w + "_")
               or ("_" + w + "_") in low for w in words)


def non_additive_kind(column: str | None) -> str | None:
    """'coordinate' | 'identifier' | 'year' | None."""
    if not column:
        return None
    if _word(column, _COORD_WORDS):
        return "coordinate"
    if _ID_NAME.search(str(column)):
        return "identifier"
    if _word(column, _YEAR_WORDS):
        return "year"
    return None


def is_quantity(column: str | None) -> bool:
    """False for a column that must never enter a measure pool."""
    return non_additive_kind(column) is None


# Aggregations that do arithmetic on the values -- meaningless for each kind.
# Mirrors ARITHMETIC in frontend/src/lib/semanticGuard.ts.
ARITHMETIC: dict[str, frozenset[str]] = {
    "coordinate": frozenset({"sum"}),
    "identifier": frozenset({"sum", "avg", "mean", "average", "median", "stddev", "variance"}),
    "year": frozenset({"sum", "avg", "mean", "average"}),
}

#: The aggregation that DOES mean something for each kind: the one-click fix.
SAFE_AGGREGATION: dict[str, str] = {"coordinate": "avg", "identifier": "countd", "year": "max"}
_SAFE_WORD = {"avg": "Average", "countd": "Distinct count", "max": "Maximum"}


def aggregation_refusal(column: str | None, aggregation: str | None) -> str | None:
    """Why (column, aggregation) is refused, or None when it means something.

    The same sentences the builder shows as a warning, so a reader of a
    refused widget and its author read one explanation."""
    kind = non_additive_kind(column)
    agg = str(aggregation or "").lower()
    if not kind or agg not in ARITHMETIC[kind]:
        return None
    what = "Summing" if agg == "sum" else "Averaging"
    fix = (f"Use {_SAFE_WORD[SAFE_AGGREGATION[kind]]} instead, or -- if {column} really is "
           f"a quantity -- mark it as a measure in the field list.")
    if kind == "coordinate":
        return f"{what} {column} adds up map coordinates -- the result is not a place. {fix}"
    if kind == "identifier":
        return f"{what} {column} does arithmetic on identifiers -- the result means nothing. {fix}"
    return f"{what} {column} adds up years -- the result is not a year. {fix}"


def config_refusal(config: dict, column_meta: dict | None, *,
                   sums_by_default: bool) -> dict | None:
    """The first measure in a widget config that is aggregated meaninglessly.

    Returns {"column", "aggregation", "safe", "message"} or None. A column the
    author marked as a measure (column_meta role) is their explicit statement
    that it IS a quantity, and passes. `sums_by_default` is whether this widget
    type sums a measure that names no aggregation (a bar does; a histogram
    plots raw values and does not)."""
    meta = column_meta or {}
    roles = config.get("roles") if isinstance(config.get("roles"), dict) else {}
    pairs = [
        (config.get("measure") or roles.get("measure"),
         config.get("aggregation") or config.get("agg")),
        (config.get("measure2") or roles.get("measure2"),
         config.get("aggregation2") or config.get("aggregation") or config.get("agg")),
    ]
    for column, agg in pairs:
        if not isinstance(column, str) or not column:
            continue
        if agg is None and not sums_by_default:
            continue
        agg = agg or "sum"
        if (meta.get(column) or {}).get("role") == "measure":
            continue
        message = aggregation_refusal(column, agg)
        if message:
            kind = non_additive_kind(column)
            return {"column": column, "aggregation": str(agg).lower(),
                    "safe": SAFE_AGGREGATION[kind], "message": message}
    return None
